lib/authz/serialized_project_update_workflow_test.go

```diff
@@ -0,0 +1,1260 @@
+				Running: serializedProjectUpdateWorkflowPhaseRunningData{
+					TotalTasks: 1,
+					RunningTask: serializedProjectUpdateRunningTask{
+						Task: serializedProjectUpdateTask{
+							DomainService: "svc-a",
+							SerializedProjectUpdateTask: SerializedProjectUpdateTask{
+								Priority: 4,
+								Params:   map[string]string{"name": "bar"},
+							},
+						},
+					},
+					RemainingTasks: []serializedProjectUpdateTask{},
+				},
+			})
+
+		executor.OnTaskComplete(instance, ev)
+		continuing := instance.AssertContinuing()
+
+		state := SerializedProjectUpdateWorkflowState{}
+		continuing.GetPayload(&state)
+
+		require.Equal(t, domainServices, state.DomainServices)
+		require.Equal(t, projectUpdateID, state.ProjectUpdateID)
+		require.Equal(t, SerializedProjectUpdateWorkflowPhaseCanceling, state.Phase)
+		require.Equal(t, runID, state.Running.RunningTask.ID)
+
+		require.Equal(t, 0, state.Running.CompletedTasks)
+		require.Equal(t, 1, state.Running.TotalTasks)
+
+		taskEnqueued := continuing.AssertTaskEnqueued(cereal.NewTaskName("project-update-cancel-task/svc-a"))
+		taskEnqueued.AssertCount(1)
+		taskParams := SerializedProjectUpdateCancelTaskTaskParams{}
+		taskEnqueued.Tasks[0].GetParameters(&taskParams)
+		require.Equal(t, runID, taskParams.ID)
+		require.Equal(t, projectUpdateID, taskParams.ProjectUpdateID)
+	})
+
+	t.Run("fails if task failed", func(t *testing.T) {
+		domainServices := []string{"svc-a"}
+		projectUpdateID := "project-update-id"
+
+		executor := NewSerializedWorkflowExecutor()
+		params := SerializedProjectUpdateWorkflowParams{
+			ProjectUpdateID: projectUpdateID,
+			DomainServices:  domainServices,
+		}
+
+		runID := SerializedProjectUpdateTaskID("run-id-1")
+		runTaskResult := SerializedProjectUpdateRunTaskTaskResult{
+			ID: runID,
+			Status: SerializedProjectUpdateTaskStatus{
+				State: SerializedProjectUpdateTaskFailed,
+				Error: "failed",
+			},
+		}
+		tr := cerealtest.NewTaskResult(t).WithResult(runTaskResult)
+		ev := cerealtest.NewTaskCompleteEvent(
+			cereal.NewTaskName("project-update-run-task/svc-a"),
+			tr,
+		)
+
+		instance := cerealtest.
+			NewWorkflowInstance(t, "serialized").
+			WithParameters(params).
+			WithPayload(SerializedProjectUpdateWorkflowState{
+				DomainServices:  domainServices,
+				ProjectUpdateID: projectUpdateID,
+				Phase:           SerializedProjectUpdateWorkflowPhaseRunning,
+				Running: serializedProjectUpdateWorkflowPhaseRunningData{
+					TotalTasks: 1,
+					RunningTask: serializedProjectUpdateRunningTask{
+						Task: serializedProjectUpdateTask{
+							DomainService: "svc-a",
+							SerializedProjectUpdateTask: SerializedProjectUpdateTask{
+								Priority: 4,
+								Params:   map[string]string{"name": "bar"},
+							},
+						},
+					},
+					RemainingTasks: []serializedProjectUpdateTask{},
+				},
+			})
+
+		executor.OnTaskComplete(instance, ev)
+		instance.AssertFailed()
+	})
+
+	t.Run("fails if error returned", func(t *testing.T) {
```

Comment:
what's the difference btwn this test and the one above? is it where the error occurs?

Reply:
Just 2 different ways of indicating error. The run task is expected to return a status. You would probably return an error in the task status if you ran a thing synchronously and it ran to completion with a failing result. You would return an error if for example you couldn't talk to elasticsearch.

I'm thinking one day, we could retry this type of error as we don't know what happened, but we wouldn't retry the previous one